src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -0,0 +1,404 @@
+                // (6,18): error CS9200: List patterns may not be used for a value of type 'int'.
+                //         _ = 0 is {0};
+                Diagnostic(ErrorCode.ERR_UnsupportedTypeForListPattern, "{0}").WithArguments("int").WithLocation(6, 18),
+                // (7,18): error CS9202: Length patterns may not be used for a value of type 'int'.
+                //         _ = 0 is [0];
+                Diagnostic(ErrorCode.ERR_UnsupportedTypeForLengthPattern, "[0]").WithArguments("int").WithLocation(7, 18),
+                // (8,18): error CS9202: Length patterns may not be used for a value of type 'int'.
+                //         _ = 0 is [0]{0};
+                Diagnostic(ErrorCode.ERR_UnsupportedTypeForLengthPattern, "[0]").WithArguments("int").WithLocation(8, 18),
+                // (9,18): error CS9200: List patterns may not be used for a value of type 'int'.
+                //         _ = 0 is {..0};
+                Diagnostic(ErrorCode.ERR_UnsupportedTypeForListPattern, "{..0}").WithArguments("int").WithLocation(9, 18),
+                // (9,19): error CS9201: Slice patterns may not be used for a value of type 'int'.
+                //         _ = 0 is {..0};
+                Diagnostic(ErrorCode.ERR_UnsupportedTypeForSlicePattern, "..0").WithArguments("int").WithLocation(9, 19));
+        }
+
+        [Fact]
+        public void ListPattern_ObsoleteMembers()
+        {
+            var source = @"
+using System;
+class X
+{
+    [Obsolete(""error"", error: true)]
+    public int Slice(int i, int j) => 0;
+    [Obsolete(""error"", error: true)]
+    public int this[int i] => 0;
+    [Obsolete(""error"", error: true)]
+    public int Count => 0;
+
+    public void M()
+    {
+        _ = this is {0};
+        _ = this is [0];
+        _ = this is [1]{0};
+        _ = this is {..0};
+    } 
+}
+";
+            // PROTOTYPE(list-patterns) Missing error for [0]: https://github.com/dotnet/roslyn/issues/53418
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularWithListPatterns);
+            compilation.VerifyDiagnostics(
+                // (14,21): error CS0619: 'X.this[int]' is obsolete: 'error'
+                //         _ = this is {0};
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "{0}").WithArguments("X.this[int]", "error").WithLocation(14, 21),
+                // (16,24): error CS0619: 'X.this[int]' is obsolete: 'error'
+                //         _ = this is [1]{0};
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "{0}").WithArguments("X.this[int]", "error").WithLocation(16, 24),
+                // (17,21): error CS0619: 'X.this[int]' is obsolete: 'error'
+                //         _ = this is {..0};
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "{..0}").WithArguments("X.this[int]", "error").WithLocation(17, 21));
+        }
+
+        [Fact]
+        public void SlicePattern_Misplaced()
```

Comment:
I'll think more about tests tomorrow. Here's some tests to add. 
Nullability and semantic model ones can be PROTOTYPE'd if needed.

- LangVersion check for all three kinds of patterns (checks for `IDS_FeatureListPattern`, `IDS_FeatureSlicePattern` and `IDS_FeatureLengthPattern` are currently missing)
- verify order of evaluation with `expectedOutput:`

For slice pattern:
- `..var x` and verify the type of `var` with `GetTypeInfo`/`GetDeclaredSymbol` and nullable state, also in error case
- slice pattern as top-level pattern:  `is ..`, `e switch { .. => ... }` and `switch (e) { case ..: ...}`  (disallowed)
- slice pattern inside length pattern `[..]`
- `{ 1, .., 2, .., 3 }` 

For length and list patterns:
- verify that pattern includes a null check on input: `maybeNull is { 1, 2, 3 }` and `maybeNull is [3]`
- static or inaccessible indexing or slicing symbol (coverage for cases in `PerformPatternIndexerLookup`)
- verify that a switch on a tuple with different nested length checks share the computed `lengthTemp` and null check: `(e1, e2) switch { ([1], _) => ..., ([2], _) => ... }` (I expect that just falls out of the existing processing of DAG)
- pattern inside length pattern `[< 3]`
- capture length with `[var length]` and verify type of `var` with `GetTypeInfo`/`GetDeclaredSymbol`, also in error case

Reply:
I think I should check IDS_FeatureListPattern in the parser among others. At that point, we've already determined if we have a list or property pattern clause.

Reply:
Some more scenarios from the spec:

> The length_pattern_clause must be in agreement with the inferred length from the list_pattern_clause (if any), e.g. [0] {1} is an error.
> However, [1] {} is not an error due to the length mismatch, rather, {} would be always parsed as an empty property_pattern_clause. We may want to add a warning for it so it would not be confused that way.
> If the type is an array_type, the length_pattern_clause is disambiguated so that int[] [0] would match an empty integer array.

I think the subsumption checking is not part of this PR. I'll add a note to teste plan to not forget.

Reply:
It should *just work* because it's already being handled in the parser. I'll verify while addressing the reminder of comments. Thanks.

Reply:
:memo: Note to self: tests to be added

- missing types (Index, Range)
- missing members (indexer, Slice, Length/Count, Array.Length, substring)
- obsolete members (indexer, Slice, Length/Count individually)
- static indexer (possible in VB)
- setter-only indexers
- inaccessible indexers
- overridden indexers
- non-readonly members in readonly members
- ref return
- params, default arguments codegen
- slice from start/end
- use-site diagnostics
- order of evaluation
- narrowed types
- impossible match
- subsumption checking
- symbols

I think there are a few "unlikely" scenarios involving inheritance or indexer signatures that is not possible in C#. I'll look into other tests to see if there's anything we should cover.

Reply:
Reviewed the list above. Thanks for addressing those test suggestions.

Did you address the following two?
- static or inaccessible indexing or slicing symbol: (`ListPattern_MemberLookup_StaticIndexer` and `ListPattern_MemberLookup_ErrorCases` covers static and inaccessible indexing symbol.) Missing similar scenarios for slicing symbol.
- pattern inside length pattern [< 3]

Reply:
>pattern inside length pattern [< 3].

See `LengthPattern`

>(mark as PROTOTYPE if doesn'tt warn)

That's an impossible match. We produce error for that. Only redundant patterns won't warn (https://github.com/dotnet/roslyn/issues/52674)